File: 3-Algorithms/Searching/3-BFS/bfs.js
// Breadth First Search (BFS) definition: traverses a graph in a breadthward motion and uses a queue to remember to get the next vertex to start a search, when a dead end occurs in any iteration.
// Time complexity: O(V+E) where V is number of vertices in the graph and E is number of edges in the graph.
// Space Complexity: O(V)

class Graph {
  constructor() {
    this.adjacencyList = new Map();
  }

  addNode(node) {
    this.adjacencyList.set(node, []);
  }

  addEdge(node1, node2) {
    this.adjacencyList.get(node1).push(node2);
    this.adjacencyList.get(node2).push(node1); // For undirected graph
  }

  bfs(startNode) {
    const visited = new Set();
    const queue = [];

    queue.push(startNode);
    visited.add(startNode);

    while (queue.length > 0) {
      const currentNode = queue.shift();
      console.log(`Visited Node: ${currentNode}`);

      const neighbors = this.adjacencyList.get(currentNode);
      for (const neighbor of neighbors) {
        if (!visited.has(neighbor)) {
          queue.push(neighbor);
          visited.add(neighbor);
        }
      }
    }
  }
}

// Example Usage:
const graph = new Graph();

graph.addNode("A");
graph.addNode("B");
graph.addNode("C");
graph.addNode("D");
graph.addNode("E");

graph.addEdge("A", "B");
graph.addEdge("A", "C");
graph.addEdge("B", "D");
graph.addEdge("C", "E");

console.log("BFS Traversal:");
graph.bfs("A");

File: 3-Algorithms/Searching/3-BFS/bfs.py
# BFS algorithm definition: it is a graph search algorithm that begins at the root node and explores all the neighboring nodes.
# Then for each of those nearest nodes, it explores their unexplored neighbor nodes, and so on, until it finds the goal.
# Time complexity: O(V+E) where V is the number of vertices in the graph and E is the number of edges in the graph.
# Space complexity: O(V) where V is the number of vertices in the graph.

class Node:
    def __init__(self, name):
        self.name = name
        self.adjacencyList = []  # represents the neighbors of the node
        self.visited = False
        self.predecessor = None


def bfs(startNode):
    queue = [startNode]

    while queue:
        actualNode = queue.pop(0)  # pop the first element of the list
        actualNode.visited = True
        print("%s " % actualNode.name)

        for n in actualNode.adjacencyList:  # iterate over the neighbors of the node
            if not n.visited:
                n.visited = True
                queue.append(n)


if __name__ == "__main__":
    # create nodes
    node1 = Node("A")
    node2 = Node("B")
    node3 = Node("C")
    node4 = Node("D")
    node5 = Node("E")

    # create edges
    node1.adjacencyList.append(node2)
    node1.adjacencyList.append(node3)
    node2.adjacencyList.append(node4)
    node4.adjacencyList.append(node5)

    # run bfs
    bfs(node1)
